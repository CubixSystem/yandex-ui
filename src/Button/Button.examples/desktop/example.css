.Showcase-Item .Button2 {
    margin-right: 10px;
    margin-bottom: 20px;
}
