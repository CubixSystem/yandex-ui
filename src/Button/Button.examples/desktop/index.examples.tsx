import { withKnobs } from '@storybook/addon-knobs';

import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export * from './view.examples';
export * from './pin.examples';
export * from './size.examples';
export * from './theme.examples';
export * from './width.examples';
export * from './playground.examples';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: [withKnobs, ...createDecorators({ scope: 'desktop' })],
    parameters,
};
