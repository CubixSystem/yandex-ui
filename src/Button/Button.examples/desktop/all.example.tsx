import React from 'react';
import { Button } from '@yandex-lego/components/Button/desktop/bundle';
import { Icon } from '@yandex-lego/components/Icon/Icon.bundle/desktop';

import './example.css';
const view = ['default', 'action', 'pseudo', 'link', 'clear', 'raised'];
const rPins = ['round-round', 'round-clear', 'clear-round', 'round-brick', 'brick-round'];
const cPins = ['circle-circle', 'circle-clear', 'clear-circle', 'circle-brick', 'brick-circle'];
const qPins = ['brick-brick', 'brick-clear', 'clear-brick'];

export const All = () => (
    <>
        {view.map((view: any) => {
            return (
                <>
                    <div className="Showcase-Item">
                        <Button view={view} size="l" title={`Button view ${view} size l`}>
                            Button
                        </Button>
                        <Button view={view} size="m" title={`Button view ${view} size m`}>
                            Button
                        </Button>
                        <Button view={view} size="s" title={`Button view ${view} size s`}>
                            Button
                        </Button>
                        <Button view={view} size="s" disabled title={`Button view ${view} disabled`}>
                            Button
                        </Button>
                        <Button view={view} size="s" checked title={`Button view ${view} checked`}>
                            Button
                        </Button>
                        <Button view={view} size="s" progress title={`Button view ${view} progress`}>
                            Button
                        </Button>
                        <Button
                            view={view}
                            size="s"
                            iconLeft={(className: string) => (
                                <Icon size="s" type="arrow" direction="left" className={className} />
                            )}
                            iconRight={(className: string) => (
                                <Icon size="s" type="arrow" direction="left" className={className} />
                            )}
                            title={`Button with icon view ${view} size s`}
                        >
                            Button
                        </Button>
                        <Button
                            view={view}
                            size="s"
                            icon={(className: string) => (
                                <Icon size="s" type="arrow" direction="left" className={className} />
                            )}
                            title={`Button with icon view ${view} size s`}
                        />
                    </div>
                </>
            );
        })}

        {view.map((view: any) => {
            return (
                <>
                    <div className="Showcase-Item">
                        {rPins.map((pin: any) => (
                            <>
                                <Button title={`Button view ${view} pin ${pin}`} key={pin} pin={pin} view={view} size="m">
                                    Button
                                </Button>
                                <Button
                                    title={`Button view ${view} pin ${pin}`}
                                    icon={(className: string) => (
                                        <Icon size="s" type="arrow" direction="left" className={className} />
                                    )} key={pin} pin={pin} view={view} size="m" />
                            </>
                        ))}
                        <br />
                        {cPins.map((pin: any) => (
                            <>
                                <Button title={`Button view ${view} pin ${pin}`} key={pin} pin={pin} view={view} size="m">
                                    Button
                                </Button>
                                <Button
                                    title={`Button view ${view} pin ${pin}`}
                                    icon={(className: string) => (
                                        <Icon size="s" type="arrow" direction="left" className={className} />
                                    )} key={pin} pin={pin} view={view} size="m" />
                            </>
                        ))}
                        <br />
                        {qPins.map((pin: any) => (
                            <>
                                <Button title={`Button view ${view} pin ${pin}`} key={pin} pin={pin} view={view} size="m">
                                    Button
                                </Button>
                                <Button
                                    title={`Button view ${view} pin ${pin}`}
                                    icon={(className: string) => (
                                        <Icon size="s" type="arrow" direction="left" className={className} />
                                    )} key={pin} pin={pin} view={view} size="m" />
                            </>
                        ))}
                        <Button title={`Button view ${view} pin clear-clear`} pin="clear-clear" view={view} size="m">
                            Button
                        </Button>
                    </div>
                </>
            );
        })}
    </>
);
