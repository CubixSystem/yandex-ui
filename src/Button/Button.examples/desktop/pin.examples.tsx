import React from 'react';
import { Button } from '@yandex-lego/components/Button/desktop/bundle';

export const Pin = () => {
    const rPins = ['round-round', 'round-clear', 'clear-round', 'round-brick', 'brick-round'];
    const cPins = ['circle-circle', 'circle-clear', 'clear-circle', 'circle-brick', 'brick-circle'];
    const qPins = ['brick-brick', 'brick-clear', 'clear-brick'];

    return (
        <>
            <p>Кнопки со скругленными уголками</p>
            {rPins.map((pin: any) => (
                <Button key={pin} pin={pin} view="default" size="m">
                    Button
                </Button>
            ))}
            <p>Кнопки с круглыми границами</p>
            {cPins.map((pin: any) => (
                <Button key={pin} pin={pin} view="default" size="m">
                    Button
                </Button>
            ))}
            <p>Кнопки с прямоугольными уголками</p>
            {qPins.map((pin: any) => (
                <Button key={pin} pin={pin} view="default" size="m">
                    Button
                </Button>
            ))}
            <p>Кнопка без боковых границ</p>
            <Button pin="clear-clear" view="default" size="m">
                Button
            </Button>
        </>
    );
};

Pin.story = {
    name: 'pin',
};
