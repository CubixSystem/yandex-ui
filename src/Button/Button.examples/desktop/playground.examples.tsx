import React from 'react';
import { select, boolean, text } from '@storybook/addon-knobs';

import { Button } from '../../Button.bundle/desktop';
import { Icon } from '../../../Icon/Icon.bundle/desktop';

import { cnTheme } from '../../../Theme';
import { presets, presetsKeys } from '../../../Theme/presets';

export const Playground = () => {
    const preset = select('theme-preset', presetsKeys, 'default');
    const children = text('children', 'Button');
    const view = select('view', ['default', 'action', 'pseudo', 'link', 'clear', 'raised', ''], 'default') as any;
    const size = select('size', ['s', 'm', 'l'], 'm') as any;
    const theme =
        view === ''
            ? (select('theme', ['action', 'clear', 'normal', 'pseudo', 'raised', 'link'], 'normal') as any)
            : null;
    const checked = boolean('checked', false) as any;
    const disabled = boolean('disabled', false) as any;
    const progress = boolean('progress', false) as any;
    const icon = boolean('icon', false);
    const iconLeft = icon ? null : boolean('iconLeft', false);
    const iconRight = icon ? null : boolean('iconRight', false);
    const iconType =
        icon || iconLeft || iconRight
            ? (select('icon type', ['arrow', 'close', 'cross', 'cross-websearch', 'filter'], 'arrow') as any)
            : null;
    const direction =
        iconType === 'arrow' ? (select('direction', ['left', 'right', 'top', 'bottom'], 'bottom') as any) : null;

    return (
        <div className={cnTheme(presets[preset])}>
            <Button
                theme={theme}
                size={size}
                view={view}
                progress={progress}
                disabled={disabled}
                checked={checked}
                children={icon ? null : children}
                // @ts-ignore TODO: Исправить тип
                icon={
                    icon &&
                    ((className: string) => (
                        <Icon size={size} type={iconType} direction={direction} className={className} />
                    ))
                }
                // @ts-ignore TODO: Исправить тип
                iconLeft={
                    iconLeft &&
                    ((className: string) => (
                        <Icon size={size} type={iconType} direction={direction} className={className} />
                    ))
                }
                // @ts-ignore TODO: Исправить тип
                iconRight={
                    iconRight &&
                    ((className: string) => (
                        <Icon size={size} type={iconType} direction={direction} className={className} />
                    ))
                }
            />
        </div>
    );
};

Playground.story = {
    name: 'playground',
};
