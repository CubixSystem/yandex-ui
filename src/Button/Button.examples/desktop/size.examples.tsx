import React from 'react';

import { Button } from '../../Button.bundle/desktop';

export const Size = () => (
    <>
        <Button view="default" size="l">
            Button
        </Button>{' '}
        <Button view="default" size="m">
            Button
        </Button>{' '}
        <Button view="default" size="s">
            Button
        </Button>
    </>
);

Size.story = {
    name: 'size',
};
