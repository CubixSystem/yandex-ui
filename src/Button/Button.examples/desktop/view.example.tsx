import React from 'react';
import { Button } from '@yandex-lego/components/Button/desktop/bundle';

export const View = () => (
    <>
        <Button view="default" size="m">
            Button
        </Button>{' '}
        <Button view="action" size="m">
            Button
        </Button>{' '}
        <Button view="link" size="m">
            Button
        </Button>{' '}
        <Button view="pseudo" size="m">
            Button
        </Button>{' '}
        <Button view="clear" size="m">
            Button
        </Button>{' '}
        <Button view="raised" size="m">
            Button
        </Button>
    </>
);
