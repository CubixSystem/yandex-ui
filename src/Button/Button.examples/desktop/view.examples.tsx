import React from 'react';

import { Button } from '../../Button.bundle/desktop';

export const View = () => (
    <>
        <Button view="default" size="m">
            Button
        </Button>{' '}
        <Button view="action" size="m">
            Button
        </Button>{' '}
        <Button view="link" size="m">
            Button
        </Button>{' '}
        <Button view="pseudo" size="m">
            Button
        </Button>{' '}
        <Button view="clear" size="m">
            Button
        </Button>{' '}
        <Button view="raised" size="m">
            Button
        </Button>
    </>
);

View.story = {
    name: 'view',
};
