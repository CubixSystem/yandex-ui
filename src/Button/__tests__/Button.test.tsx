import React, { createRef, FC } from 'react';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { Keys } from '../../lib/keyboard';
import { Button as ButtonCommon, ContainerElement } from '../Button';
import { Button as ButtonDesktop } from '../desktop';
import { Button as ButtonTouchPad } from '../touch-pad';
import { Button as ButtonTouchPhone } from '../touch-phone';

const platforms = [
    ['desktop', ButtonDesktop],
    ['touch-pad', ButtonTouchPad],
    ['touch-phone', ButtonTouchPhone],
] as const;

describe.each(platforms)('Button@%s', (_platform, Button) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Button
                id="clientTemplateId"
                className="mix"
                iconLeft={(className) => <i className={className + ' iconLeft'} />}
                iconRight={(className) => <i className={className + ' iconRight'} />}
            />,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLElement>();
        render(<Button innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать ссылку на нативный DOM элемент', () => {
        const controlRef = createRef<ContainerElement>();
        render(<Button controlRef={controlRef} />);

        expect(controlRef.current).not.toBe(null);
    });

    test('должен быть установлен корректный displayName', () => {
        expect(ButtonCommon.displayName).toBe('Button2');
    });

    test('должен рендерить контент в addonAfter и addonBefore', () => {
        render(<Button addonAfter={<i data-testid="after" />} addonBefore={<i data-testid="before" />} />);

        expect(screen.getByTestId('after')).toBeInTheDocument();
        expect(screen.getByTestId('before')).toBeInTheDocument();
    });

    test('при checked=true должен быть установлен класс Button2_checked и prop aria-pressed', () => {
        render(<Button checked autoComplete="off" />);

        expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'true');
        expect(screen.getByRole('button')).toHaveClass('Button2_checked');
    });

    test('autocomplete прокидывается в корневой элемент компонента', () => {
        render(<Button autoComplete="on" />);

        expect(screen.getByRole('button')).toHaveAttribute('autoComplete', 'on');
    });

    test('тег компонента зависит от prop tag', () => {
        render(<Button as="span" data-testid="element" />);

        expect(screen.getByTestId('element').tagName).toBe('SPAN');
    });

    test('children рендерится в DOM', () => {
        render(<Button>Привет</Button>);

        expect(screen.getByRole('button')).toHaveTextContent('Привет');
    });

    test('иконка рендерится', () => {
        render(<Button icon={() => <i data-testid="icon" />} />);

        expect(screen.getByTestId('icon')).toBeInTheDocument();
    });

    test('при совместной установке icon и iconLeft рендерится только iconLeft', () => {
        render(<Button iconLeft={() => <i data-testid="icon-left" />} icon={() => <i data-testid="icon" />} />);

        expect(screen.getByTestId('icon-left')).toBeInTheDocument();
        expect(screen.queryByTestId('icon')).not.toBeInTheDocument();
    });

    test('при установке progress=true создается класс Button2_progress и значение aria-busy устанавливается в true', () => {
        render(<Button progress />);

        expect(screen.getByRole('button')).toHaveClass('Button2_progress');
        expect(screen.getByRole('button')).toHaveAttribute('aria-busy', 'true');
        expect(screen.getByRole('button')).toHaveAttribute('disabled');
    });

    test('при установке prop disapled в DOM-элементе должны быть props aria-disabled=true и disabled', () => {
        render(<Button disabled />);

        expect(screen.getByRole('button')).toHaveAttribute('aria-disabled', 'true');
        expect(screen.getByRole('button')).toHaveAttribute('disabled');
    });

    test('className пробрасывается в корневой элемент', () => {
        render(<Button className="testClassName" />);

        expect(screen.getByRole('button')).toHaveClass('testClassName');
    });

    test('при совместной установке checked, progress, className стили корректны', () => {
        render(<Button checked progress className="testClassName" onClick={() => null} />);

        expect(screen.getByRole('button')).toHaveClass('testClassName');
        expect(screen.getByRole('button')).toHaveClass('Button2_checked');
        expect(screen.getByRole('button')).toHaveClass('Button2_progress');
    });

    test('при совместной установке props innerRef и controlRef они должны оба указывать на root DOM-элемент', async () => {
        const controlRef = createRef<ContainerElement>();
        const innerRef = createRef<ContainerElement>();
        render(<Button controlRef={controlRef} innerRef={innerRef} />);

        expect(controlRef.current).toBe(innerRef.current);
        expect(controlRef.current).not.toBe(null);
    });

    test('при нажатии на кнопку вызывается функция, переданная в prop onClick', () => {
        const onClickFn = jest.fn();
        render(<Button onClick={onClickFn} />);

        fireEvent.click(screen.getByRole('button'));
        expect(onClickFn).toHaveBeenCalledTimes(1);

        fireEvent.click(screen.getByRole('button'));
        expect(onClickFn).toHaveBeenCalledTimes(2);
    });

    test('при нажатии на кнопку вызывается функция, переданная в prop onMouseDown', () => {
        const onMouseDownFn = jest.fn();
        render(<Button onMouseDown={onMouseDownFn} />);

        fireEvent.mouseDown(screen.getByRole('button'));
        expect(onMouseDownFn).toHaveBeenCalledTimes(1);

        fireEvent.mouseDown(screen.getByRole('button'));
        expect(onMouseDownFn).toHaveBeenCalledTimes(2);
    });

    test('при нажатии на кнопку вызывается функция, переданная в prop onKeyUp', () => {
        // TODO: при переходе на 17 реакт `e.persist` убрать
        const onKeyUpFn = jest.fn((e) => e.persist());
        render(<Button onKeyUp={onKeyUpFn} />);

        fireEvent.keyUp(screen.getByRole('button'), { keyCode: Keys.DELETE });

        expect(onKeyUpFn).toHaveBeenCalledTimes(1);
        expect(onKeyUpFn).toHaveBeenCalledWith(expect.objectContaining({ keyCode: Keys.DELETE }));

        fireEvent.keyUp(screen.getByRole('button'), { keyCode: Keys.BACKSPACE });

        expect(onKeyUpFn).toHaveBeenCalledTimes(2);
        expect(onKeyUpFn).toHaveBeenCalledWith(expect.objectContaining({ keyCode: Keys.BACKSPACE }));
    });

    test('при нажатии на кнопку вызывается функция, переданная в prop onKeyDown', () => {
        // TODO: при переходе на 17 реакт `e.persist` убрать
        const onKeyDownFn = jest.fn((e) => e.persist());
        render(<Button onKeyDown={onKeyDownFn} />);

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.DELETE });

        expect(onKeyDownFn).toHaveBeenCalledTimes(1);
        expect(onKeyDownFn).toHaveBeenCalledWith(expect.objectContaining({ keyCode: Keys.DELETE }));

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.BACKSPACE });

        expect(onKeyDownFn).toHaveBeenCalledTimes(2);
        expect(onKeyDownFn).toHaveBeenCalledWith(expect.objectContaining({ keyCode: Keys.BACKSPACE }));
    });

    test('при нажатии на кнопку, которая передана в pressKeys должен добавляться класс Button2_pressed', () => {
        render(<Button pressKeys={[Keys.ESC, Keys.TAB]} />);

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.ESC });

        expect(screen.getByRole('button')).toHaveClass('Button2_pressed');

        fireEvent.keyUp(screen.getByRole('button'), { keyCode: Keys.TAB });

        expect(screen.getByRole('button')).not.toHaveClass('Button2_pressed');
    });

    test('при нажатии на кнопку, которая передана в preventKeys должно предотвращаться default поведение', () => {
        const preventDefault = jest.spyOn(KeyboardEvent.prototype, 'preventDefault');
        render(<Button prvntKeys={[Keys.ESC, Keys.TAB]} pressKeys={[Keys.ESC, Keys.TAB]} />);

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.ESC });

        expect(preventDefault).toBeCalledTimes(1);

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.TAB });

        expect(preventDefault).toBeCalledTimes(2);

        preventDefault.mockReset();
    });

    test('должен устанавливать фокус в интерактивный элемент с различными компонентами', () => {
        const Link: FC<any> = ({ innerRef, ...props }) => <a href="#a" ref={innerRef} {...props} />;
        const { setProps } = render(<Button />);

        fireEvent.click(screen.getByRole('button'));

        expect(screen.getByRole('button')).toHaveFocus();

        setProps({ as: Link });
        fireEvent.click(screen.getByRole('link'));

        expect(screen.getByRole('link')).toHaveFocus();
    });

    test('при потере фокуса кнопка должна терять класс Button2_pressed', () => {
        render(<Button />);

        fireEvent.keyDown(screen.getByRole('button'), { keyCode: Keys.SPACE });

        expect(screen.getByRole('button')).toHaveClass('Button2_pressed');

        fireEvent.blur(screen.getByRole('button'));

        expect(screen.getByRole('button')).not.toHaveClass('Button2_pressed');
    });

    test('при при потере фокуса вызывается функция, переданная в prop onBlur', () => {
        const onBlurFn = jest.fn();
        render(<Button onBlur={onBlurFn} />);

        fireEvent.blur(screen.getByRole('button'));

        expect(onBlurFn).toHaveBeenCalledTimes(1);
    });

    test('должен вернуть шаблон компонента c <Text/> при children={0}', () => {
        render(<Button children={0} />);

        expect(screen.getByRole('button')).toHaveTextContent('0');
    });

    test('должен вернуть шаблон компонента без <Text/> при children={true}', () => {
        render(<Button children />);

        expect(screen.getByRole('button')).toBeEmptyDOMElement();
    });

    test('должен вернуть шаблон компонента с иконками без класса стороны при children={true}', () => {
        render(
            <Button
                children
                iconLeft={(className) => <i data-testid="icon-left" className={className} />}
                iconRight={(className) => <i data-testid="icon-right" className={className} />}
            />,
        );

        expect(screen.getByTestId('icon-left')).not.toHaveClass('Button2-Icon_side_left');
        expect(screen.getByTestId('icon-left')).not.toHaveClass('Button2-Icon_side_right');
    });
});
