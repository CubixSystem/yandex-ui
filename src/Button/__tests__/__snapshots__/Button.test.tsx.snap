// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<button
  autocomplete="off"
  class="Button2 mix"
  id="clientTemplateId"
  type="button"
>
  <i
    class="Button2-Icon iconLeft"
  />
  <i
    class="Button2-Icon iconRight"
  />
</button>
`;

exports[`Button@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<button
  autocomplete="off"
  class="Button2 mix"
  id="clientTemplateId"
  type="button"
>
  <i
    class="Button2-Icon iconLeft"
  />
  <i
    class="Button2-Icon iconRight"
  />
</button>
`;

exports[`Button@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<button
  autocomplete="off"
  class="Button2 mix"
  id="clientTemplateId"
  type="button"
>
  <i
    class="Button2-Icon iconLeft"
  />
  <i
    class="Button2-Icon iconRight"
  />
</button>
`;
